File: app/src/main/java/org/neidhardt/dynamicsoundboard/soundmanagement/views/SoundSettingsDialog.java
package org.neidhardt.dynamicsoundboard.soundmanagement.views;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.FragmentManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import de.greenrobot.event.EventBus;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData;
import org.neidhardt.dynamicsoundboard.dao.SoundSheet;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.soundactivity.SoundActivity;
import org.neidhardt.dynamicsoundboard.soundcontrol.SoundSheetFragment;
import org.neidhardt.dynamicsoundboard.soundmanagement.ServiceManagerFragment;
import org.neidhardt.dynamicsoundboard.soundmanagement.events.AddNewSoundEvent;
import org.neidhardt.dynamicsoundboard.views.edittext.CustomEditText;
import org.neidhardt.dynamicsoundboard.views.spinner.CustomSpinner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * File created by eric.neidhardt on 23.02.2015.
 */
public class SoundSettingsDialog extends SoundSettingsBaseDialog implements View.OnClickListener, CompoundButton.OnCheckedChangeListener
{
	private static final String TAG = SoundSettingsDialog.class.getName();

	private CustomEditText soundName;
	private CustomEditText soundSheetName;
	private CustomSpinner soundSheetSpinner;
	private CheckBox addNewSoundSheet;

	private int indexOfCurrentFragment = -1;

	public static void showInstance(FragmentManager manager, MediaPlayerData playerData)
	{
		SoundSettingsDialog dialog = new SoundSettingsDialog();
		addArguments(dialog, playerData.getPlayerId(), playerData.getFragmentTag());
		dialog.show(manager, TAG);
	}

	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState)
	{
		@SuppressLint("InflateParams") View view = this.getActivity().getLayoutInflater().inflate(R.layout.dialog_sound_settings_layout, null);

		this.soundName = (CustomEditText)view.findViewById(R.id.et_name_file);
		this.soundSheetName = (CustomEditText)view.findViewById(R.id.et_name_new_sound_sheet);
		this.soundSheetSpinner = (CustomSpinner)view.findViewById(R.id.s_sound_sheets);
		this.addNewSoundSheet = (CheckBox)view.findViewById(R.id.cb_add_new_sound_sheet);

		this.addNewSoundSheet.setOnCheckedChangeListener(this);

		view.findViewById(R.id.b_cancel).setOnClickListener(this);
		view.findViewById(R.id.b_ok).setOnClickListener(this);

		this.soundName.setText(this.player.getMediaPlayerData().getLabel());

		this.setAvailableSoundSheets();

		this.soundSheetName.setText(this.soundSheetsDataUtil.getSuggestedName());
		this.soundSheetName.setVisibility(View.GONE);

		AlertDialog.Builder builder = new AlertDialog.Builder(this.getActivity());
		builder.setView(view);

		return builder.create();
	}

	private void setAvailableSoundSheets()
	{
		List<SoundSheet> soundSheets = this.soundSheetsDataAccess.getSoundSheets();
		ArrayList<String> labels = new ArrayList<>();
		for (int i = 0; i < soundSheets.size(); i++)
		{
			if (soundSheets.get(i).getFragmentTag().equals(this.fragmentTag))
				this.indexOfCurrentFragment = i;
			labels.add(soundSheets.get(i).getLabel());
		}
		if (this.indexOfCurrentFragment == -1)
			throw new IllegalStateException(TAG + " Current fragment of sound " + this.player.getMediaPlayerData() + " is not found in list of sound sheets " + soundSheets);

		this.soundSheetSpinner.setItems(labels);
		this.soundSheetSpinner.setSelectedItem(this.indexOfCurrentFragment);
	}

	@Override
	public void onCheckedChanged(CompoundButton buttonView, boolean isChecked)
	{
		if (isChecked)
		{
			this.soundSheetSpinner.setVisibility(View.GONE);
			this.soundSheetName.setVisibility(View.VISIBLE);
		}
		else
		{
			this.soundSheetName.setVisibility(View.GONE);
			this.soundSheetSpinner.setVisibility(View.VISIBLE);
		}
	}

	@Override
	public void onClick(View v)
	{
		switch (v.getId())
		{
			case R.id.b_cancel:
				this.dismiss();
				break;
			case R.id.b_ok:
				boolean hasLabelChanged = !this.player.getMediaPlayerData().getLabel().equals(this.soundName.getDisplayedText());
				this.deliverResult();
				this.dismiss();
				if (hasLabelChanged)
					RenameSoundFileDialog.showInstance(this.getFragmentManager(), this.player.getMediaPlayerData());
				break;
		}
	}

	private void deliverResult()
	{
		String soundLabel = this.soundName.getDisplayedText();
		int indexOfSelectedSoundSheet = this.soundSheetSpinner.getSelectedItemPosition();

		boolean addNewSoundSheet = this.addNewSoundSheet.isChecked();
		boolean hasSoundSheetChanged = addNewSoundSheet || indexOfSelectedSoundSheet != this.indexOfCurrentFragment;

		SoundSheetFragment soundSheetFragment = this.getSoundSheetFragment(this.fragmentTag);
		if (!hasSoundSheetChanged)
		{
			this.player.getMediaPlayerData().setLabel(soundLabel);
			soundSheetFragment.notifyDataSetChanged();
		}
		else
		{
			ServiceManagerFragment serviceManagerFragment = this.getServiceManagerFragment();
			serviceManagerFragment.getSoundService().removeSounds(Collections.singletonList(this.player));
			soundSheetFragment.notifyDataSetChanged();

			Uri uri = Uri.parse(this.player.getMediaPlayerData().getUri());

			MediaPlayerData mediaPlayerData;

			if (addNewSoundSheet)
			{
				String soundSheetName = this.soundSheetName.getDisplayedText();
				SoundSheet soundSheet = this.soundSheetsDataUtil.getNewSoundSheet(soundSheetName);
				String fragmentTag = this.soundSheetsDataStorage.addOrUpdateSoundSheet(soundSheet);

				mediaPlayerData = EnhancedMediaPlayer.getMediaPlayerData(fragmentTag, uri, soundLabel);
			}
			else
				mediaPlayerData = EnhancedMediaPlayer.getMediaPlayerData(this.fragmentTag, uri, soundLabel);

			EventBus.getDefault().post(new AddNewSoundEvent(mediaPlayerData, false));
		}
	}

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/soundmanagement/dialog/SoundSettingsDialog.kt
package org.neidhardt.dynamicsoundboard.soundmanagement.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.view.View
import android.widget.CheckBox
import kotlinx.android.synthetic.main.dialog_sound_settings_layout.view.*
import org.greenrobot.eventbus.EventBus
import org.neidhardt.android_utils.views.CustomEditText
import org.neidhardt.android_utils.views.SimpleSpinner
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.manager.NewSoundSheetManager
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerFactory
import org.neidhardt.dynamicsoundboard.persistance.model.NewMediaPlayerData
import org.neidhardt.dynamicsoundboard.persistance.model.NewSoundSheet
import org.neidhardt.dynamicsoundboard.soundmanagement.events.SoundChangedEvent
import org.neidhardt.utils.letThis
import java.util.*
import kotlin.properties.Delegates

/**
 * File created by eric.neidhardt on 23.02.2015.
 */
class SoundSettingsDialog : SoundSettingsBaseDialog() {

	override var fragmentTag: String by Delegates.notNull<String>()
	override var player: MediaPlayerController by Delegates.notNull<MediaPlayerController>()
	override var soundSheet: NewSoundSheet? = null

	private var soundName: CustomEditText? = null
	private var soundSheetName: CustomEditText? = null
	private var soundSheetSpinner: SimpleSpinner? = null
	private var addNewSoundSheet: CheckBox? = null

	private var indexOfCurrentFragment = -1

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
		@SuppressLint("InflateParams") val view = this.activity.layoutInflater.inflate(R.layout.dialog_sound_settings_layout, null)

		this.soundName = view.et_name_file.letThis {
			it.text = this.player.mediaPlayerData.label
		}
		this.soundSheetName = view.et_name_new_sound_sheet.letThis {
			it.text = this.soundSheetManager.suggestedName
			it.visibility = View.GONE
		}
		this.soundSheetSpinner = view.s_sound_sheets?.letThis {
			this.setAvailableSoundSheets(it)
		}
		this.addNewSoundSheet = view.cb_add_new_sound_sheet.letThis {
			it.setOnCheckedChangeListener { view,isChecked ->
				this.soundSheetSpinner?.visibility = if (isChecked) View.GONE else View.VISIBLE
				this.soundSheetName?.visibility = if (isChecked) View.VISIBLE else View.GONE
			}
		}

		return AlertDialog.Builder(context).apply {
			this.setTitle(R.string.dialog_sound_settings_title)
			this.setView(view)
			this.setPositiveButton(R.string.dialog_save, { dialogInterface, i ->
				val hasLabelChanged = player.mediaPlayerData.label != soundName!!.displayedText
				deliverResult()
				dismiss()
				if (hasLabelChanged)
					RenameSoundFileDialog.show(fragmentManager, player.mediaPlayerData)
			})
			this.setNegativeButton(R.string.dialog_cancel, { dialogInterface, i -> dismiss()})
		}.create()
	}

	private fun setAvailableSoundSheets(spinner: SimpleSpinner) {
		val soundSheets = this.soundSheetManager.soundSheets
		val labels = ArrayList<String>()
		for (i in soundSheets.indices) {
			if (soundSheets[i].fragmentTag == this.fragmentTag)
				this.indexOfCurrentFragment = i
			labels.add(soundSheets[i].label)
		}
		if (this.indexOfCurrentFragment == -1)
			throw IllegalStateException(TAG + " Current fragment of sound " + this.player.mediaPlayerData + " is not found in list of sound sheets " + soundSheets)

		spinner.items = labels
		spinner.setSelection(this.indexOfCurrentFragment)
	}

	private fun deliverResult() {
		val soundLabel = this.soundName!!.displayedText
		val indexOfSelectedSoundSheet = this.soundSheetSpinner!!.selectedItemPosition

		val addNewSoundSheet = this.addNewSoundSheet!!.isChecked
		val hasSoundSheetChanged = addNewSoundSheet || indexOfSelectedSoundSheet != this.indexOfCurrentFragment

		if (!hasSoundSheetChanged) {
			this.player.mediaPlayerData.label = soundLabel
			EventBus.getDefault().post(SoundChangedEvent(this.player))
		} else {
			if (soundSheet != null)
				this.soundManager.remove(soundSheet!!, listOf(this.player))
			else
				this.playlistManager.remove(listOf(this.player))


			val uri = Uri.parse(this.player.mediaPlayerData.uri)

			val mediaPlayerData: NewMediaPlayerData
			if (addNewSoundSheet) {
				val soundSheetName = this.soundSheetName!!.displayedText
				val fragmentTag = NewSoundSheetManager.getNewFragmentTagForLabel(soundSheetName)
				val soundSheet = NewSoundSheet().apply {
					this.fragmentTag = fragmentTag
					this.label = soundSheetName
				}
				this.soundSheetManager.add(soundSheet)

				mediaPlayerData = MediaPlayerFactory.getNewMediaPlayerData(fragmentTag, uri, soundLabel)
			} else {
				val fragmentTag = this.soundSheetManager.soundSheets[indexOfSelectedSoundSheet].fragmentTag
				mediaPlayerData = MediaPlayerFactory.getNewMediaPlayerData(fragmentTag, uri, soundLabel)
			}

			if (this.soundSheet != null)
				this.soundManager.add(soundSheet!!, mediaPlayerData)
			else
				this.playlistManager.add(mediaPlayerData)
		}
	}

	companion object {
		private val TAG = SoundSettingsDialog::class.java.name

		fun showInstance(manager: FragmentManager, playerData: NewMediaPlayerData) {
			val dialog = SoundSettingsDialog()
			addArguments(dialog, playerData.playerId, playerData.fragmentTag)
			dialog.show(manager, TAG)
		}
	}

}
